rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,826 @@
+
+- AWS account nickname (e.g. `mbta-ctd-test`, `mbta-ctd-main`, or preferably a more succinct version that eliminates the `mbta-` prefix)
+- Team name (prefixed with `team-`)
+- Environment context (e.g. `dev`, `staging`, `prod`) - this overlaps with AWS account nickname, so we may want to discourage the use of both at the same time
+- Component descriptor of the type/category of resources the module is intended to contain, if it's one of multiple modules for the particular provider/account/context (e.g. `base`, `bootstrap`, `restricted`)
+- Nothing (in cases where one module encompasses all of a provider's configuration)
+
+A side-note regarding the AWS account nicknames: Our main AWS account is currently named "mbtacomsupport", but we are hoping to [rename it](https://docs.aws.amazon.com/accounts/latest/reference/manage-acct-alias.html) soon. We toyed with the name "mbta-ctd-prod", but since the account also contains development/staging resources&mdash;which can't reasonably be provisioned into their own separate "development/staging" AWS account&mdash;we've opted to recommend the name "mbta-ctd-main". That name is open to debate, however.
+
+#### Proposed Module Names
+
+With all the above in mind, the following are the proposed names for all existing and future root modules. (Note that the naming convention for child modules is different, and is covered in the [module versioning][guide-module-versioning] section.)
+
+Root modules associated with the "CTD Production" AWS account that should be renamed:
+
+- `bootstrap` → `aws-ctd-main-bootstrap`
+- `global` → `aws-ctd-main-base`
+- `dev` → multiple modules:
+  - move application development/staging resources to team-based modules:
+    - `aws-ctd-main-team-dataplatform`
+    - `aws-ctd-main-team-glides`
+    - `aws-ctd-main-team-lamp`
+    - `aws-ctd-main-team-pass-programs`
+    - `aws-ctd-main-team-screens`
+    - `aws-ctd-main-team-trc`
+    - `aws-ctd-main-team-website`
+  - move base dev resources to `aws-ctd-main-base`
+- `onprem_dev` → `aws-onprem-dev`
+- `onprem_prod` → `aws-onprem-prod`
+- `prod` → `aws-ctd-main-apps` (to remain as a single module for now)
+- `restricted` → `aws-ctd-main-restricted`
+
+Root modules to remain as-is:
+
+- `scalr`
+
+New root modules for "mbta-ctd-test" AWS account:
+
+- `aws-ctd-test-base`
+- `aws-ctd-test-bootstrap`
+- `aws-ctd-test-restricted`
+
+Possible future root modules:
+
+- `github`
+- `sentry`
+- `splunk`
+
+## 2. Module Testing, Detailed
+[reference-module-testing]: #2-module-testing-detailed
+
+As noted in the [guide-level explanation][guide-module-testing], we will set up testing for each application and infrastructure component child module by creating a `test/` module within each module's directory. The test module will define any dependent resources required by the module, and then call the module itself. It could even call other application modules if needed.
```

Comment:
i would think that you have something more like

```
terraform/aws-mbta-main-team/
    variables.tf
    outputs.tf
    main.tf
    staging/
        variables.tf
        staging.tf
    dev/
        variables.tf
        dev.tf
        dev-blue.tf
    test/
        variables.tf
        main.tf
```

Reply:
The module testing workflow wouldn't apply to root modules, only child modules. Since your `aws-mbta-main-team` module is a root module, it won't contain any subdirectories, only `.tf` files that reference child modules such as `app-something`. It's your `app-something` child module that would contain its own testing module within it. All testing would be on the individual child module, and changes would then bubble up to the root module once tested, merged (and optionally versioned).

Does that make sense?